web/src/lib/managers/timeline-manager/timeline-manager.svelte.ts

```diff
@@ -0,0 +1,547 @@
+      return;
+    }
+    await this.initTask.reset();
+    await this.#init(options);
+    this.#updateViewportGeometry(false);
+  }
+
+  async #init(options: TimelineManagerOptions) {
+    this.isInitialized = false;
+    this.months = [];
+    this.albumAssets.clear();
+    await this.initTask.execute(async () => {
+      this.#options = options;
+      await this.#initializeMonthGroups();
+    }, true);
+  }
+
+  public destroy() {
+    this.disconnect();
+    this.isInitialized = false;
+  }
+
+  async updateViewport(viewport: Viewport) {
+    if (viewport.height === 0 && viewport.width === 0) {
+      return;
+    }
+
+    if (this.viewportHeight === viewport.height && this.viewportWidth === viewport.width) {
+      return;
+    }
+
+    if (!this.initTask.executed) {
+      await (this.initTask.loading ? this.initTask.waitUntilCompletion() : this.#init(this.#options));
+    }
+
+    const changedWidth = viewport.width !== this.viewportWidth;
+    this.viewportHeight = viewport.height;
+    this.viewportWidth = viewport.width;
+    this.#updateViewportGeometry(changedWidth);
+  }
+
+  #updateViewportGeometry(changedWidth: boolean) {
+    if (!this.isInitialized) {
+      return;
+    }
+    if (this.viewportWidth === 0 || this.viewportHeight === 0) {
+      return;
+    }
+    for (const month of this.months) {
+      updateGeometry(this, month, { invalidateHeight: changedWidth });
+    }
+    this.updateIntersections();
+    this.#createScrubberMonths();
+  }
+
+  #createScrubberMonths() {
+    this.scrubberMonths = this.months.map((month) => ({
+      assetCount: month.assetsCount,
+      year: month.yearMonth.year,
+      month: month.yearMonth.month,
+      title: month.monthGroupTitle,
+      height: month.height,
+    }));
+    this.scrubberTimelineHeight = this.timelineHeight;
+  }
+
+  createLayoutOptions() {
+    const viewportWidth = this.viewportWidth;
+
+    return {
+      spacing: 2,
+      heightTolerance: 0.15,
+      rowHeight: this.#rowHeight,
+      rowWidth: Math.floor(viewportWidth),
+    };
+  }
+
+  async loadMonthGroup(yearMonth: TimelinePlainYearMonth, options?: { cancelable: boolean }): Promise<void> {
+    let cancelable = true;
+    if (options) {
+      cancelable = options.cancelable;
+    }
+    const monthGroup = getMonthGroupByDate(this, yearMonth);
+    if (!monthGroup) {
+      return;
+    }
+
+    if (monthGroup.loader?.executed) {
+      return;
+    }
+
+    const result = await monthGroup.loader?.execute(async (signal: AbortSignal) => {
+      await loadFromTimeBuckets(this, monthGroup, this.#options, signal);
+    }, cancelable);
+    if (result === 'LOADED') {
+      updateIntersectionMonthGroup(this, monthGroup);
+    }
+  }
+
+  addAssets(assets: TimelineAsset[]) {
+    const assetsToUpdate: TimelineAsset[] = [];
+
+    for (const asset of assets) {
+      if (this.isExcluded(asset)) {
+        continue;
+      }
+      assetsToUpdate.push(asset);
+    }
```

Comment:
Could arguably also be a `const assetsToUpdate = assets.filter((asset) => this.isExcluded(asset))`